src/ripple/app/ledger/impl/LedgerConsensusImp.cpp

```diff
@@ -0,0 +1,1876 @@
+            << threshConsensus << " Pos:" << closeTime;
+    }
+
+    if (!changes &&
+            ((closeTime != roundCloseTime (mOurPosition->getCloseTime (), mCloseResolution))
+            || mOurPosition->isStale (ourCutoff)))
+    {
+        // close time changed or our position is stale
+        ourPosition = mAcquired[mOurPosition->getCurrentHash ()]
+            ->snapShot (true);
+        assert (ourPosition);
+        changes = true; // We pretend our position changed to force
+    }                   //   a new proposal
+
+    if (changes)
+    {
+        uint256 newHash = ourPosition->getHash ();
+        WriteLog (lsINFO, LedgerConsensus)
+            << "Position change: CTime " << closeTime
+            << ", tx " << newHash;
+
+        if (mOurPosition->changePosition (newHash, closeTime))
+        {
+            if (mProposing)
+                propose ();
+
+            mapCompleteInternal (newHash, ourPosition, false);
+        }
+    }
+}
+
+void LedgerConsensusImp::playbackProposals ()
+{
+    for (auto const& it: getApp().getOPs ().peekStoredProposals ())
+    {
+        bool relay = false;
+        for (auto const& proposal : it.second)
+        {
+            if (proposal->isPrevLedger (mPrevLedgerHash))
+                relay = peerPosition (proposal);
+
+            if (relay)
+            {
+                WriteLog (lsWARNING, LedgerConsensus)
+                    << "We should do delayed relay of this proposal,"
+                    << " but we cannot";
+            }
+        }
+    }
+}
+
+void LedgerConsensusImp::closeLedger ()
+{
+    checkOurValidation ();
+    state_ = State::establish;
+    mConsensusStartTime
+        = std::chrono::steady_clock::now ();
+    mCloseTime = getApp().getOPs ().getCloseTimeNC ();
+    consensus_.setLastCloseTime (mCloseTime);
+    statusChange (protocol::neCLOSING_LEDGER, *mPreviousLedger);
+    getApp().getLedgerMaster().applyHeldTransactions ();
+    takeInitialPosition (*getApp().getLedgerMaster ().getCurrentLedger ());
+}
+
+void LedgerConsensusImp::checkOurValidation ()
+{
+    // This only covers some cases - Fix for the case where we can't ever
+    // acquire the consensus ledger
+    if (!mHaveCorrectLCL || !mValPublic.isSet ()
+        || !mValPrivate.isSet ()
+        || getApp().getOPs ().isNeedNetworkLedger ())
+    {
+        return;
+    }
+
+    auto lastValidation = consensus_.getLastValidation ();
+
+    if (lastValidation)
+    {
+        if (lastValidation->getFieldU32 (sfLedgerSequence)
+            == mPreviousLedger->getLedgerSeq ())
+        {
+            return;
+        }
+        if (lastValidation->getLedgerHash () == mPrevLedgerHash)
+            return;
+    }
+
+    uint256 signingHash;
+    auto v = std::make_shared<STValidation> (mPreviousLedger->getHash (),
+        consensus_.validationTimestamp (), mValPublic, false);
+    addLoad(v);
+    v->setTrusted ();
+    signingHash = v->sign (mValPrivate);
```

Comment:
I think you can postpone the construction of `signingHash` to here, similar to the very nice change you made around line 1040.  You might even be able to make it `const`.